### PR TITLE
Support Accelerator OOM Error (#121200)

### LINKED ISSUE #121200: Make 'torch._C._OutOfMemoryError' more genenral for other devices except CUDA.
### 📚 The doc issue

I've found the docs for 'torch._C._OutOfMemoryError' in 'torch/csrc/Exceptions.cpp' like the following:
```cpp
...
ASSERT_TRUE(
      THPException_OutOfMemoryError = PyErr_NewExceptionWithDoc(
          "torch.cuda.OutOfMemoryError",
          "Exception raised when CUDA is out of memory",
          PyExc_RuntimeError,
          nullptr));
  ASSERT_TRUE(
      PyModule_AddObject(
          module, "_OutOfMemoryError", THPException_OutOfMemoryError) == 0);
...
```
There are hard code for CUDA. Since more devices were added to PyTorch backend(using privateuse1) and they may need this Error too except for CUDA. I was wondering if there is any need to change this Error.(For example change the error name to "torch.OutOfMemoryError" and doc to "Exception raised when device is out of memory").
Please let me know if I should make a PR.

### Suggest a potential alternative/fix

```cpp
...
ASSERT_TRUE(
      THPException_OutOfMemoryError = PyErr_NewExceptionWithDoc(
          "torch.OutOfMemoryError",
          "Exception raised when device is out of memory",
          PyExc_RuntimeError,
          nullptr));
  ASSERT_TRUE(
      PyModule_AddObject(
          module, "_OutOfMemoryError", THPException_OutOfMemoryError) == 0);
...
```

cc @albanD

### PR DESCRIPTION
**Pull request by PHLens**
Fixes #121200 
This PR introduces AcceleratorOutOfMemoryError for all privateuse1 backend. For python, there is a PyError object which will be set only when privateuse1 is registered. All privateuse1 backend then can use this error for memory errors. Maybe more error types in the future.

cc @albanD

### CODE REVIEW

File: torch/csrc/Module.cpp [hunk 1]
```diff
@@ -616,6 +634,9 @@ static PyObject* THModule_rename_privateuse1_backend(
       THPUtils_typename(arg));
   const std::string backend_name = THPUtils_unpackString(arg);
   c10::register_privateuse1_backend(backend_name);
+  if (c10::is_privateuse1_backend_registered()) {
```
Comment (albanD) [inline torch/csrc/Module.cpp:637]: We should always register this.

File: torch/csrc/Module.cpp [hunk 2]
```diff
@@ -606,6 +607,23 @@ PyObject* THModule_getCppBacktrace(PyObject* _unused, PyObject* args) {
   END_HANDLE_TH_ERRORS
 }
 
+// Set Accelerator Python Errors.
+// Only used when privateuse1 is registered.
+static void setAcceleratorPyError() {
+  auto device_type = at::getAccelerator(true);
```
Comment (albanD) [inline torch/csrc/Module.cpp:613]: At initialization time here, privateuse1 is most likely not registered yet. So this will fail?

**Review by albanD — COMMENTED** [review]
cc @guangyey I guess you want this one to work for you as well.
Can you help review this PR to make sure it works properly for XPU as well (as in your allocator should be able to throw AcceleratorOutOfMemoryError directly and it will be nicely translated to python)?

File: torch/csrc/Module.cpp (hunk 2 above)
Comment (PHLens) [reply torch/csrc/Module.cpp:613]: This is only called inside the `THModule_rename_privateuse1_backend` func. At initialization time, this func is not called, It is only called when the privateuse1 try to register. And the `setAcceleratorPyError` will be called just after `c10::register_privateuse1_backend`, so I think this is ok.

File: torch/csrc/Module.cpp (hunk 1 above)
Comment (PHLens) [reply torch/csrc/Module.cpp:637]: ok, I'll delete this.

**Comment by guangyey** [comment]
> cc @guangyey I guess you want this one to work for you as well. Can you help review this PR to make sure it works properly for XPU as well (as in your allocator should be able to throw AcceleratorOutOfMemoryError directly and it will be nicely translated to python)?

Thanks, @albanD. In my opinion, `OutOfMemoryError` is general enough. The cuda-bias code exists in [`THPException_OutOfMemoryError`](https://github.com/pytorch/pytorch/blob/014f91a9d9f94ac9a7f0711600240d7cd7f69844/torch/csrc/Exceptions.cpp#L62-L66) and is hard to generalize it as the exception name can't be modified after its initialization. Otherwise, we need to change it at compilation time.
Is it reasonable to only change `THPException_OutOfMemoryError ` to
```cpp
ASSERT_TRUE(
      THPException_OutOfMemoryError = PyErr_NewExceptionWithDoc(
          "torch.OutOfMemoryError",
          "Exception raised when device is out of memory",
          PyExc_RuntimeError,
          nullptr));
```
Then, as usual, expose `torch._C._OutOfMemoryError` to `torch.xxx.OutOfMemoryError` so that each backend uses `except torch.xxx.OutOfMemoryError` to catch the OOM exception. like cuda:
`OutOfMemoryError = torch._C._OutOfMemoryError` defined in torch.cuda.\_\_init_\_.py

**Comment by guangyey** [comment]
> > they can use except torch.xxx.OutOfMemoryError to catch the OOM exception(for backward compatibility)
> 
> What about having a single `torch.OutOfMemoryError()` that is used by all device and we just have `torch.cuda....` be an alias to this for BC reasons only?

Yes, agreed.

**Comment by PHLens** [comment]
So, we only need to change the docs of `THPException_OutOfMemoryError` for other devices. Is it true? @guangyey @albanD

**Comment by guangyey** [comment]
> So, we only need to change the docs of `THPException_OutOfMemoryError` for other devices. Is it true? @guangyey @albanD

I think yes. refine the doc and name about [THPException_OutOfMemoryError ](https://github.com/pytorch/pytorch/blob/270dd99180112125eab5f91baf32c48682ec3c8e/torch/csrc/Exceptions.cpp#L62).

**Comment by PHLens** [comment]
I've updated the changes, pls review it, thanks. @guangyey @albanD

**Review by guangyey — DISMISSED** [review]
LGTM.

**Comment by PHLens** [comment]
It seems I should add a doc for the new `torch.OutOfMemoryError` class, would you pls tell me where I should put the docs? @guangyey @albanD

**Comment by guangyey** [comment]
> It seems I should add a doc for the new `torch.OutOfMemoryError` class, would you pls tell me where I should put the docs? @guangyey @albanD

I think [here](https://github.com/pytorch/pytorch/blob/main/docs/source/torch.rst) is a good place.

**Comment by PHLens** [comment]
> > It seems I should add a doc for the new `torch.OutOfMemoryError` class, would you pls tell me where I should put the docs? @guangyey @albanD
> 
> I think [here](https://github.com/pytorch/pytorch/blob/main/docs/source/torch.rst) is a good place.

Thanks, but I've found that most of the APIs have no descriptions in the `torch.rst`, e.g. https://github.com/pytorch/pytorch/blob/ec00daf4f1b09b830a2dfc300a859b5d6f9f39e5/docs/source/torch.rst?plain=1#L25, however I find the description on the website https://pytorch.org/docs/main/torch.html#tensors. I've searched the codespace and found no such description, so I don't know how to add the description about `torch.OutOfMemoryError`, can you give me some advice? Thanks. @guangyey

**Comment by guangyey** [comment]
> > > It seems I should add a doc for the new `torch.OutOfMemoryError` class, would you pls tell me where I should put the docs? @guangyey @albanD
> > 
> > 
> > I think [here](https://github.com/pytorch/pytorch/blob/main/docs/source/torch.rst) is a good place.
> 
> Thanks, but I've found that most of the APIs have no descriptions in the `torch.rst`, e.g.
> 
> https://github.com/pytorch/pytorch/blob/ec00daf4f1b09b830a2dfc300a859b5d6f9f39e5/docs/source/torch.rst?plain=1#L25
> 
> , however I find the description on the website https://pytorch.org/docs/main/torch.html#tensors. I've searched the codespace and found no such description, so I don't know how to add the description about `torch.OutOfMemoryError`, can you give me some advice? Thanks. @guangyey

refer to https://github.com/pytorch/pytorch/blob/ec00daf4f1b09b830a2dfc300a859b5d6f9f39e5/docs/source/cuda.rst?plain=1#L39

**Comment by PHLens** [comment]
I've pushed the doc changes. Do you mean to change like this? @guangyey

File: torch/csrc/Exceptions.cpp [hunk 3]
```diff
@@ -60,8 +60,8 @@ could not be completed because the input matrix is singular.",
   // NOLINTNEXTLINE(bugprone-assignment-in-if-condition)
   ASSERT_TRUE(
       THPException_OutOfMemoryError = PyErr_NewExceptionWithDoc(
-          "torch.cuda.OutOfMemoryError",
-          "Exception raised when CUDA is out of memory",
+          "torch.OutOfMemoryError",
+          "Exception raised when device is out of memory",
           PyExc_RuntimeError,
```
Comment (guangyey) [inline torch/csrc/Exceptions.cpp:65]: Another method is to rename `torch._C._OutOfMemoryError` to `torch.OutOfMemoryError` as below:
```cpp
  PyTypeObject* type = (PyTypeObject*)THPException_OutOfMemoryError;
  type->tp_name = "torch.OutOfMemoryError";

  ASSERT_TRUE(
      PyModule_AddObject(
          module, "OutOfMemoryError", THPException_OutOfMemoryError) == 0);
```
Then you can remove `OutOfMemoryError`'s description from the `torch.rst`. Meanwhile, change `_OutOfMemoryError` to `OutOfMemoryError` in `torch/_C/__init__.pyi.in` and `torch/cuda/\_\_init\_\_.py`
I think this is better.

Comment (PHLens) [reply torch/csrc/Exceptions.cpp:65]: ok，updated.

Comment (guangyey) [reply torch/csrc/Exceptions.cpp:65]: Thanks for your update, please fix the CI failure.

Comment (PHLens) [reply torch/csrc/Exceptions.cpp:65]: > Thanks for your update, please fix the CI failure.

Done. All tests have passed now.

**Review by albanD — DISMISSED** [review]
Looks good!
Thanks for the PR and your help with review @guangyey

**Comment by atalman** [comment]
This is merged: https://github.com/pytorch/pytorch/commit/9aba918bd83c5524cf19672a7aeec96faa99997a closing

**Pull request closed without merging**